### PR TITLE
🐛 Fix jpeg compression issue

### LINKED ISSUE #161: 🐛 Resilience of indexing process
**Describe the bug**
Pupyl fails to index images. I have tried with different sets of images, but it always fails. After importing and scanning all images, a `FileIsNotImage` exception is raised.

**To Reproduce**
```
import os
from pathlib import Path
from pupyl.search import PupylImageSearch

p = PupylImageSearch(os.path.join(Path.home(), 'pupyl'))
p.index('https://github.com/policratus/pupyl/raw/main/samples/images.tar.xz')

```
Running the above code gives the following output:
```
...
🕔 Scanning images: 594 items.
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
FileNotFoundError: 

The above exception was the direct cause of the following exception:

IndexError                                Traceback (most recent call last)
File ~\miniconda3\lib\site-packages\pupyl\search.py:195, in PupylImageSearch.index(self, uri, **kwargs)
    194 try:
--> 195     futures = {
    196         executor.submit(
    197             self.extractor.extract_save,
    198             self.image_database.mount_file_name(
    199                 rank, extension='npy'
    200             ),
    201             self.image_database.load_image_metadata(
    202                 rank, filtered=['internal_path']
    203             )['internal_path']
    204         ): rank
    205         for rank in ranks
    206     }
    208     _ = [
    209         *self.extractor.progress(
    210             concurrent.futures.as_completed(futures),
   (...)
    213         )
    214     ]

File ~\miniconda3\lib\site-packages\pupyl\search.py:201, in <dictcomp>(.0)
    194 try:
    195     futures = {
    196         executor.submit(
    197             self.extractor.extract_save,
    198             self.image_database.mount_file_name(
    199                 rank, extension='npy'
    200             ),
--> 201             self.image_database.load_image_metadata(
    202                 rank, filtered=['internal_path']
    203             )['internal_path']
    204         ): rank
    205         for rank in ranks
    206     }
    208     _ = [
    209         *self.extractor.progress(
    210             concurrent.futures.as_completed(futures),
   (...)
    213         )
    214     ]

File ~\miniconda3\lib\site-packages\pupyl\storage\database.py:274, in ImageDatabase.load_image_metadata(self, index, **kwargs)
    273 except FileNotFoundError:
--> 274     raise IndexError(
    275         f'Index not found: {index}'
    276     ) from FileNotFoundError

IndexError: Index not found: 227

The above exception was the direct cause of the following exception:

FileIsNotImage                            Traceback (most recent call last)
Input In [168], in <cell line: 10>()
      6 from pupyl.search import PupylImageSearch
      8 p = PupylImageSearch('/content/pupyl')
---> 10 p.index('https://github.com/policratus/pupyl/raw/main/samples/images.tar.xz')

File ~\miniconda3\lib\site-packages\pupyl\search.py:217, in PupylImageSearch.index(self, uri, **kwargs)
    208         _ = [
    209             *self.extractor.progress(
    210                 concurrent.futures.as_completed(futures),
   (...)
    213             )
    214         ]
    216     except IndexError as index_error:
--> 217         raise FileIsNotImage('Please, check your input images.') \
    218             from index_error
    220 for rank in self.extractor.progress(
    221     sorted(ranks),
    222     precise=True,
    223     message='Indexing features:'
    224 ):
    225     self.indexer.append(
    226         self.extractor.load(
    227             self.image_database.mount_file_name(
   (...)
    231         check_unique=check_unique
    232     )

FileIsNotImage: Please, check your input images.
```
**Desktop (please complete the following information):**
 - OS: Windows 10
- Python: 3.9.7
- Pupyl: 0.13.5

### PR DESCRIPTION
**Pull request by 92amartins**
This is an attempt to fix #161.

I think we can iterate over this in order to find the best design.

### CODE REVIEW

File: pupyl/duplex/image.py [hunk 1]
```diff
@@ -306,3 +308,19 @@ def compress(cls, tensor, as_tensor=False):
...
+        """Whether the underlying OS supports compression.
+
+        Returns
+        -------
+        bool:
+            True if the OS supports compression, False otherwise.
+        """
+        compression_supported = True
+
+        if platform.system() == "Windows":
```
Comment (policratus) [inline pupyl/duplex/image.py:323]:
```suggestion
        if platform.system() == 'Windows':
```

File: pupyl/duplex/image.py [hunk 2]
```diff
@@ -9,6 +9,8 @@
 from tensorflow import io as io_ops
 from tensorflow import image as image_ops
 
+import platform
```
[the remark this answers is not in the document]
Comment (92amartins) [reply pupyl/duplex/image.py:12]: Fine :+1: 

Is there a reason for `platform` sit above `base64`? I guess alphabetical order might be more intuitive here.

File: tests/test_image.py [hunk 3]
```diff
@@ -150,3 +150,7 @@ def test_get_image_bytes_to_base64():
         test_image_b64 = b64encode(image_bytes)
 
     assert test_image_b64 == ImageIO.get_image_bytes_to_base64(image_bytes)
+
+
+def test_os_supports_compression():
+    isinstance(ImageIO.os_supports_compression(), bool)
```
[the remark this answers is not in the document]
Comment (92amartins) [reply tests/test_image.py:156]: Same remark here regarding alphabetical order.

File: pupyl/duplex/image.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (92amartins) [reply pupyl/duplex/image.py]: Done c872ec7df885c3d816ade540eed7c6a3ba75cec2

File: tests/test_image.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (92amartins) [reply tests/test_image.py]: Hmm, yeah, that would work. However, I don't really like the idea of repeating the if statement in this case.

What about this solution: 9ae8b6f664dc45d7ebd0317669d7edcc106eb794

The key is the use of `patch` for "setting" OS on the fly. This allows us to test both cases (windows and others), regardless of the underlying system running the code.

`pytest.mark.parametrize` is optional in this solution, we could replace with two separate unit tests.

File: pupyl/duplex/image.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (92amartins) [reply pupyl/duplex/image.py]: Done if we follow 9ae8b6f664dc45d7ebd0317669d7edcc106eb794

File: tests/test_image.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (policratus) [reply tests/test_image.py]: For sure! Really liked the solution in [9ae8b6f](https://github.com/policratus/pupyl/commit/9ae8b6f664dc45d7ebd0317669d7edcc106eb794). I'll integrate it to the final solution.

File: pupyl/duplex/image.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (policratus) [reply pupyl/duplex/image.py]: It's more a visual decision than anything else. So, imports with higher line length goes below another with lower. But it's fine to organize this way, it's just a matter of "decoration". 😄 

An "extreme" example: https://github.com/policratus/pupyl/blob/681bd390a8e96bb9ecec1462ae27189e775f6c08/pupyl/duplex/file_io.py#L3

**Review by policratus — APPROVED** [review]
👍🏼 Good to go!

**Review by policratus — DISMISSED** [review]
👍🏼

File: pupyl/storage/database.py [hunk 4]
```diff
@@ -318,7 +320,19 @@ def insert(self, index, uri):
...
 
             if not os.path.exists(mounted_file):
-                if self.is_animated_gif(uri):
+                compression_not_supported = not self.os_supports_compression()
+                if (
+                    self.is_animated_gif(uri) or compression_not_supported
+                ):
+                    warnings.warn(
+                        termcolor.colored(
+                            f'Storing {uri} with original compression due '
```
Comment (policratus) [inline pupyl/storage/database.py:329]: 👍🏼

**Review by policratus — APPROVED** [review]
👍🏼

**Pull request merged**